### ISSUE TITLE
Idea: before after hooks

### ISSUE THREAD

**Issue by hekike** [rudijs/amqp.node-rpc-factory#7]
In my application I added a hooks factory around consumer and publisher create, but I in the publisher hooks I cannot get the correlation id.

Knowing the correlation id for me it would be escpeially useful for debugging and logging features. In this way I would be able to connect the parts in third party systems.

The concept can be used by JSON messages as well, check my example below: 

What do you think about the following concept?

``` js
function myHook (msg, props) {
  console.log(props.correlationId)
}
```

``` js
var publisherOptions = {
...
beforeHooks: [myHook1, fromJS]
afterHooks: [toJS, myHook2]
}
```

``` js
var consumerOptions = {
...
beforeHooks: [toJS, myHook3]
afterHooks: [myHook4, fromJS]
}
```

I wanted to send a pr but stucked with the broken tests.

Any chance that you can take a look for the tests with 4.x.x Node.js?
Updating the packages would be also very helpful.

Thanks,
Peter

**Comment by rudijs** [comment]
Hi Peter,

> Any chance that you can take a look for the tests with 4.x.x Node.js?

I'm already using this module in production is v4.1 - the existing tests all pass on my box here local.

I'm using v5 for my current dev work and plan to keep using that going forward.

> Updating the packages would be also very helpful.

Sure I can look into this and then release an updated patch number version.

> Knowing the correlation id

Hrmmm .. my initial thoughts on that are that the correlation ID is a private variable.

It's used within the RPC process. If a msg is sent and no reply, then by default a rejected promise is returned and the nack'd msg removed from the queue - so you'll need to re-send that msg.

If you need a unique ID for tracking the message, perhaps the JSON-RPC example code might suit you.

In the JSON-RPC example there's a UUID generated for each message before it's sent.

The JSON-RPC example is based on the www.jsonrpc.org/specification

JSON-RPC is what I use in all my use cases for this NPM module.

> What do you think about the following concept?

I'm not sure I'm a fan of this approach.

The module does one thing and returns a promise - the code, though small enough, is complicated as it is.

Adding more code, and error handling, would make the code more difficult to reason about.

If you need "life cycle" hooks in your messaging process, why not look at adding them into the promise chain?

**Comment by rudijs** [comment]
Peter,

This is the JSON-RPC example, please check it out it may suit what you require:

https://github.com/rudijs/amqp.node-rpc-factory/tree/master/examples/commandline/advanced-jsonrpc

The example uses a JSON-RPC id value of 101, but in production I use uuid() for this value.

The publisher and consumer both access this ID, it acts like a correlation id but it's at a higher level in the messaging process.

**Comment by rudijs** [comment]
Published amqp-rpc-factory@1.0.2

**Comment by hekike** [comment]
Hi Rudi,

thanks for the fast reply and for updating the package.

I just realised that the tests are ok but because the error `console.log's I thought something went wrong. Sorry, my fault. I should have checked the readme more carefully, you mentioned it with the TEST_QUIET flag as well :/

`json-rpc` with uuid sounds reasonable, I'll do it.
I can understand that you would like to keep the lib small, for the hooks I'll consider to publish a plugin for `amqp.node-rpc-factory` to `npm`.

I'll let you know when I'll be done.

Thanks,
Peter
